### PR TITLE
feat(eth-contracts): TokenLocker contract and tests

### PR DESCRIPTION
**Pull request by mkxbl**
- accomplish TokenLocker.sol 
  - without refund token feature if replay resist cell misused
- TokenLocker contract tests
  - ckb burn tx data mocked by rust code
- fix ci
  - network error when first run test
  - mandatory clean even if test fail
- change utils/waitForGas to utils/waitForReceipt

### CODE REVIEW

File: eth-contracts/contracts/TokenLocker.sol [hunk 1]
```diff
@@ -1,95 +1,99 @@
...
+    bytes32 public recipientCellTypescriptCodeHash_;
+    uint8 public recipientCellTypescriptHashType_;
 
     // txHash -> Used
     mapping(bytes32 => bool) public usedTx_;
 
     event Locked(
         address indexed token,
         address indexed sender,
+        string  indexed ckbAddress,
```
Comment (JacobDenver007) [inline eth-contracts/contracts/TokenLocker.sol:38]: ckbAddress should not be ```indexed```, if a ```string``` type is indexed, it's hash will be stored in topics, thus proof generator can not decode string from topic

Comment (mkxbl) [reply eth-contracts/contracts/TokenLocker.sol:38]: right

File: eth-contracts/contracts/TokenLocker.sol [hunk 2]
```diff
@@ -27,69 +36,64 @@ contract TokenLocker {
...
-        address mockToken = address(0);
-        address mockRecipient = address(0);
-        result = BurnResult(mockAmount, mockToken, mockRecipient);
+        recipientCellTypescriptCodeHash_ = typescriptCodeHash;
+        recipientCellTypescriptHashType_ = typescriptHashType;
     }
 
     // before lockToken, user should approve -> TokenLocker Contract with 0xffffff token
-    function lockToken(address token, uint256 amount, string memory ckbAddress) public {
+    function lockToken(address token, uint256 amount, string memory ckbAddress, string memory replayResistOutpoint) public {
```
Comment (huwenchao) [inline eth-contracts/contracts/TokenLocker.sol:57]:
```suggestion
    function lockToken(address token, uint256 amount, bytes memory ckbAddress, bytes memory replayResistOutpoint) public {
```

Is bytes better for this? If we use string, we may have to hex encode it, the memory length will be longer, which may consume more gas.

File: eth-contracts/package.json [hunk 3]
```diff
@@ -4,10 +4,10 @@
   "description": "CKB SPV proofs in Solidity",
   "scripts": {
     "compile": "hardhat compile",
-    "start-geth": "docker run -d --name geth-node --mount source=geth-dag,target=/root/.ethash -p 8545:8545 tockb/geth-priv:v1.9.23",
+    "start-geth": "docker run -d --name geth-node --mount source=geth-dag,target=/root/.ethash -p 8545:8545 tockb/geth-priv:v1.9.23 && ./wait-for-it.sh -t 0 127.0.0.1:8545 -- echo \"node is up\" && node scripts/waitForMiner.js",
     "stop-geth": "docker rm -f geth-node",
-    "pretest": "npm run start-geth && ./wait-for-it.sh -t 0 127.0.0.1:8545 -- echo \"node is up\" && node scripts/waitForMiner.js",
-    "test": "hardhat test --network geth",
+    "pretest": "npm run start-geth",
+    "test": "hardhat test --network geth || (npm run stop-geth && false)",
```
Comment (huwenchao) [inline eth-contracts/package.json:10]: Why?

```
|| (npm run stop-geth && false)
```

The posttest should be triggered automatically. And it should be ok even if the previous steps fails.

File: eth-contracts/test/data/testTokenLocker.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    "codeHash": "0x8b79673c89404c1d98874fef0441bf25d99ed5c4e96d1180d72b203966e6c964",
+    "hashType": 0
+  },
+  "lockETHTestCases": [
+    {
+      "lockAmount": "1.2",
+      "ckbAddress": "ckb1qyqwmxws34634ced7l7x7vzgklx22mpmltaq683yc6",
+      "replayResistOutpoint": "0x123456"
+    }
+  ],"lockTokenTestCases": [
```
Comment (huwenchao) [inline eth-contracts/test/data/testTokenLocker.json:12]: you can format this file by `prettier`

File: eth-contracts/test/data/testTokenLocker.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
+  "recipientCellTypescript": {
```
Comment (huwenchao) [inline eth-contracts/test/data/testTokenLocker.json:2]: This should be in different level.

```
{
  "recipientCellTypescript": {},
  "testCases": {
    ...
   }
}
```

File: eth-contracts/package.json (hunk 3 above)
Comment (mkxbl) [reply eth-contracts/package.json:10]: The posttest won't be executed if test failed according to npm, what the script do is to make posttest executed even if test failed.

When we yarn test locally, we always need execute posttest whether the test succeed or not.

File: eth-contracts/test/data/testTokenLocker.json (hunk 4 above)
Comment (mkxbl) [reply eth-contracts/test/data/testTokenLocker.json:12]: ok

File: eth-contracts/contracts/TokenLocker.sol (hunk 2 above)
Comment (mkxbl) [reply eth-contracts/contracts/TokenLocker.sol:57]: yes. Rename `ckbAddress` to `recipientLockscript`, and add `sudtExtraData` field

File: eth-contracts/contracts/TokenLocker.sol [hunk 6]
```diff
@@ -1,24 +1,33 @@
...
     using TypedMemView for bytes;
     using TypedMemView for bytes29;
-    using ViewCKB for bytes29;
+    using CKBTxView for bytes29;
     using ViewSpv for bytes29;
 
+    struct BurnResult {
+        uint256 amount;
+        address token;
+        address recipient;
```
Comment (huwenchao) [inline eth-contracts/contracts/TokenLocker.sol:24]: add fee related logic.

**Review by huwenchao — APPROVED** [review]

**Pull request merged**